### DISCUSSION TITLE
Does FLAML work for self-built neural network?

### DISCUSSION THREAD

**Discussion by hanhanwu** [microsoft/FLAML#188] [Q&A]
Hello FLAML team,

I was trying to compare the performance between FLAML and Optuna in deep learning params tuning.

Starting with a basic CNN.

This is my Optuna code, which works. Params need to tune can be put in the objective function:
![optuna_cnn](https://user-images.githubusercontent.com/4024769/132144347-3d9d25de-e85c-4765-a23a-42925289746b.PNG)

Then I was trying to figure out how to use FLAML to tune the same CNN model. I have checked the notebook here: https://github.com/microsoft/FLAML/blob/main/notebook/flaml_finetune_transformer.ipynb

And mainly defined `train_cnn()` that's similar to your `train_distilbert()` in the notebook:
![flaml_train_cnn](https://user-images.githubusercontent.com/4024769/132144451-58f0abd6-ee51-45c2-a3e7-9bd3957ad9a8.PNG)

I guess the `**config` is the search space? And here's the code tried to tune the model:
![flaml_cnn](https://user-images.githubusercontent.com/4024769/132144427-3375f11e-723b-4d10-8813-2de23159f9d0.PNG)

But this doesn't work.

If FLAML also works for self-built neural networks, is that possible for you guys to add a simple notebook example for that, so that users can follow? (If you are able to tell what's wrong in my code here, that's even better :) )
I feel pre-trained model from transformer is harder to understand since I'm not familiar with transformer and guess many other users can be similar :)

Thanks a bunch!!

**Comment by sonichi** [comment]
`num_train_epochs` is not in the search space. So it needs to be removed from `low_cost_partial_config`. Could you comment the line with `low_cost_partial_config` and try again?

**Reply by hanhanwu** [reply]
Thank you for the suggestion!

It's still showing the same error. The error is indicating something is wrong with Ray, but I have tested the same Ray code without running other code, everything worked fine. So I guess something might be wrong in my FLAML code here, but hard to locate where was wrong...
![error](https://user-images.githubusercontent.com/4024769/132259361-79f49d4a-4de6-476f-bbf1-d39ec54ce22c.PNG)

**Reply by hanhanwu** [reply]
This is also why, if you guys could add a simple deep learning notebook example, might be easier for us to follow...

**Reply by sonichi** [reply]
We have a working example under test/test_pytorch_cifar10.py. But from the error message it seems you are using windows, and I haven't been able to make ray work in windows. Ray's support for windows is experimental. Could you try commenting `ray.shutdown()`, `ray.init()` and `use_ray=True` to see if it works without ray?

**Reply by sonichi** [reply]
Added a notebook example in this PR: https://github.com/microsoft/FLAML/pull/189

**Reply by hanhanwu** [reply]
Great! Thank you very much!

**Reply by sonichi** [reply]
The PR has been merged. Is there any follow-up question?

**Reply by hanhanwu** [reply]
Yeah, I saw the notebook last week. No more question so far. Thank you very much!
